File: sock-merchant/sock-merchant.clj
(defn sockMerchant [_ colors]
  (let [color-to-total
        (reduce
         (fn [ret color]
           (if (ret color)
             (update ret color inc)
             (assoc ret color 1)))
         {} colors)
        pairs-totals
        (map
         (fn [total]
           (if (= 0 (mod total 2))
             (/ total 2)
             (/ (dec total) 2)))
         (vals color-to-total))]
    pairs-totals))

File: sock-merchant/sock-merchant.js
// https://www.hackerrank.com/challenges/sock-merchant/problem

const sockMerchant = (_, colors) => {
  const colorToTotal = colors.reduce((ret, color) => {
    ret[color] = color in ret ? ++ret[color] : 1;

    return ret;
  }, {});

  const pairsTotals = Object.values(colorToTotal).map(
    total => (total % 2 === 0 ? total : --total) / 2
  );

  const pairsTotalsSum = pairsTotals.reduce(
    (ret, pairsTotal) => (pairsTotal >= 1 ? ret + pairsTotal : ret),
    0
  );

  return pairsTotalsSum;
};
